### PR TITLE
Fix and improve OriginValidator

### PR DESCRIPTION
**Pull request by Chukak**
See issue  #929 

Added: 
* methods
    * `self.valid_origin_full(self, origin)` 
    * `self.validate_host_full(self, origin)`
    * `self.is_allowed_origin(self, origin, pattern)`
    * `self.get_origin_ports(self, origin)`
    * `self.check_pattern(self, origin, pattern)`
* attributes
    * `self.full`
    * `self.check_cert`

The ability to check only domains (by default) or to check the entire source is added. You can add both complete and incomplete sources to the list of allowed domains, but for complete verification, at least a schema and domain are required. You can also check the ssl certificate.

Also included are tests and an update for the documentation.

### CODE REVIEW

**Comment by bluetech** [comment]
By an unfortunate coincidence, I have just written [my own version](https://github.com/bluetech/channels/commits/fix-origin-validator) of this change. Since you already submitted, I'll tuck mine away, though you can take a look and see if there's something helpful there.

Thanks for working on the issue!

File: channels/security/websocket.py [hunk 1]
```diff
@@ -10,26 +12,38 @@ class OriginValidator:
...
     Validates that the incoming connection has an Origin header that
     is in an allowed list.
+
+    If full attr is True, any origin from allowed_origins must be scheme://hostname[:port].
+    Port is optional, but recommended.
+    Check_cert works only with full attr.
     """
 
-    def __init__(self, application, allowed_origins):
+    def __init__(self, application, allowed_origins, full=False, check_cert=False):
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: Can we have a better name for this than `full`?

File: channels/security/websocket.py [hunk 2]
```diff
@@ -40,9 +54,88 @@ def valid_origin(self, origin):
         # None is not allowed
         if origin is None:
             return False
+        # Get only hostname all allowed origins
+        # Only if not full: https://domain.example.com:8443 -> domain.example.com
+        #   or //domain.example.com - > domain.example.com
+        #   or .example.com -> .example.com
+        self.allowed_origins = [urlparse(pattern).hostname or urlparse("//" + pattern).hostname
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: Can you rewrite this without using `or`? Two in a row makes it harder to skim over.

File: channels/security/websocket.py [hunk 3]
```diff
@@ -40,9 +54,88 @@ def valid_origin(self, origin):
...
+                                or pattern for pattern in self.allowed_origins]
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: The current code indent style is one level and the trailing bracket on the next line - please try and match that!

File: channels/security/websocket.py [hunk 4]
```diff
@@ -10,26 +12,38 @@ class OriginValidator:
...
-                    origin_host = urlparse(header_value.decode("ascii")).hostname
+                    # Set ResultParse and origin hostname
+                    result_parse = urlparse(header_value.decode("ascii"))
+                    origin_hostname = result_parse.hostname
                 except UnicodeDecodeError:
                     pass
-        # Check to see if the origin header is valid
-        if self.valid_origin(origin_host):
+        # Check to see if the origin header is valid. Full or default check
+        valid = self.valid_origin(origin_hostname) if not self.full else self.valid_origin_full(result_parse)
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: I'd rather do this check inside a single `valid_origin` function.

File: channels/security/websocket.py [hunk 5]
```diff
@@ -40,9 +54,88 @@ def valid_origin(self, origin):
...
+        return False
+
+    def get_origin_ports(self, origin):
+        if origin.port is not None:
+            # Return tuple (origin.port, )
+            return origin.port,
+        # if origin.port doesn`t exists
+        if origin.scheme == "http":
+            # Default ports return for http
+            return 80, 8080
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: Is port 8080 actually specified as a default HTTP port anywhere?

Comment (hundredeir) [reply channels/security/websocket.py]: Yes

File: channels/security/websocket.py (hunk 1 above)
Comment (Chukak) [reply channels/security/websocket.py]: Can you suggest a better name than `` full``, maybe ``origin`` or ``scheme``. Is it necessary to leave the possibility of verifying the certificate or include it in the full check origin.

**Comment by Chukak** [comment]
I updated my code. Now, ``OriginValidator`` does not require any options. In ``allowed_origins``, you can pass only domain or origin as scheme://domain[:port] (for exapmple,  ``.domain.com``, ``*``, ``http://.domain.com``.)

File: channels/security/websocket.py [hunk 6]
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
...
+    def validate_origin(self, parsed_origin):
+        """
+        Validate the given origin for this site.
+
+        Check than the origin looks valid and matches the origin pattern in
+        specified list `` allowed_origins``. Any pattern begins with a scheme.
+        After the scheme there must be a domain. Any domain beginning with a period
+        corresponds to the domain and all its subdomains (for example, ``http://.example.com``
+        ``http://example.com`` and any subdomain).After the domain there must be a port,
+        but it can be omitted. `` * `` matches anything and anything
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: No spaces around the `*` here I think.

File: channels/security/websocket.py [hunk 7]
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
...
+
+    def validate_origin(self, parsed_origin):
+        """
+        Validate the given origin for this site.
+
+        Check than the origin looks valid and matches the origin pattern in
+        specified list `` allowed_origins``. Any pattern begins with a scheme.
+        After the scheme there must be a domain. Any domain beginning with a period
+        corresponds to the domain and all its subdomains (for example, ``http://.example.com``
+        ``http://example.com`` and any subdomain).After the domain there must be a port,
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: Space after the `.` needed

File: channels/security/websocket.py [hunk 8]
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
...
+        but port is optional.
+
+        Returns `` True`` for a valid host, `` False`` otherwise.
+        """
+        return any(
+            pattern == "*" or self.is_allowed_origin(parsed_origin, pattern)
+            for pattern in self.allowed_origins
+        )
+
+    def is_allowed_origin(self, parsed_origin, pattern):
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: I would call this `match_allowed_origin` so it's obviously different from `validate_origin`

File: channels/security/websocket.py [hunk 9]
```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
...
+            pattern_hostname = urlparse("//" + pattern).hostname or pattern
+            return is_same_domain(parsed_origin.hostname, pattern_hostname)
+        # Get origin.port or default ports for origin or None
+        origin_port = self.get_origin_port(parsed_origin)
+        # Get pattern.port or default ports for pattern or None
+        pattern_port = self.get_origin_port(parsed_pattern)
+        # Compares hostname, scheme, ports of pattern and origin
+        if parsed_pattern.scheme == parsed_origin.scheme and origin_port == pattern_port \
+                and is_same_domain(parsed_origin.hostname, parsed_pattern.hostname):
+            # Check ssl security certificate
```
Comment (andrewgodwin) [inline channels/security/websocket.py]: I do not think we should try to check the certificate in this code at all, and definitely not by default.

**Comment by Chukak** [comment]
I updated the code in accordance with your recommendations. Yes, you were right, `` * .example.com`` is a backwards incompatible change, I added tests for this.

File: docs/topics/security.rst [hunk 10]
```diff
@@ -37,10 +39,17 @@ valid domains as the second argument::
...
                 ])
             ),
-            ["goodsite.com", "*.goodsite.com"],
+            [".goodsite.com", "http://.goodsite.com:80", "http://other.site.com"],
         ),
     })
 
+Note: If you want to resolve any domains, just add ``["*"]`` as second argument.
+
+``OriginValidator`` can accept the ``check_cert`` as fourth argument.
```
Comment (andrewgodwin) [inline docs/topics/security.rst]: `check_cert` is no longer there

File: docs/topics/security.rst [hunk 11]
```diff
@@ -25,7 +25,9 @@ ASGI middlewares it contains, ``OriginValidator`` and
 ``OriginValidator`` lets you restrict the valid options for the ``Origin``
 header that is sent with every WebSocket to say where it comes from. Just wrap
 it around your WebSocket application code like this, and pass it a list of
-valid domains as the second argument::
+valid domains as the second argument. You can pass only domain (for example,
```
Comment (andrewgodwin) [inline docs/topics/security.rst]: "only a single domain"

File: docs/topics/security.rst [hunk 12]
```diff
@@ -25,7 +25,9 @@ ASGI middlewares it contains, ``OriginValidator`` and
...
+``.allowed-domain.com``) and full origin, than must be scheme://domain[:port]
```
Comment (andrewgodwin) [inline docs/topics/security.rst]: "and full origin, than must be" -> "or a full origin, in the format" (and make the format code-style formatting)

File: docs/topics/security.rst [hunk 13]
```diff
@@ -37,10 +39,17 @@ valid domains as the second argument::
                     ...
                 ])
             ),
-            ["goodsite.com", "*.goodsite.com"],
+            [".goodsite.com", "http://.goodsite.com:80", "http://other.site.com"],
         ),
     })
 
+Note: If you want to resolve any domains, just add ``["*"]`` as second argument.
```
Comment (andrewgodwin) [inline docs/topics/security.rst]: "any domains, just add * as second argument" -> "any domain, then use the origin `*`"

**Comment by Chukak** [comment]
I updated the docs for websockets.

**Comment by andrewgodwin** [comment]
Thanks so much!

**Pull request merged**